lib/portage/sync/revision_history.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2024 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+import json
+import os
+from typing import Optional
+
+import portage
+from portage.locks import lockfile, unlockfile
+from portage.repository.config import RepoConfig
+from portage.util.path import first_existing
+
+_HISTORY_LIMIT = 25
+
+
+def get_repo_revision_history(
+    eroot: str, repos: Optional[list[RepoConfig]] = None
+) -> dict[str, list[str]]:
+    """
+    Get revision history of synced repos. Returns a dict that maps
+    a repo name to list of revisions in descending order by time.
+    If a change is detected and the current process has permission
+    to update the repo_revisions file, then the file will be updated
+    with any newly detected revisions.
+
+    This functions detects revisions which are not yet visible to the
+    current process due to the sync-rcu option.
+
+    @param eroot: EROOT to query
+    @type eroot: string
+    @param repos: list of RepoConfig instances to check for new revisions
+    @type repos: list
+    @rtype: dict
+    @return: mapping of repo name to list of revisions in descending
+             order by time
```

Comment:
Ultimately, when binhosts clients need to checkout a specific source repo revision, we don't want that to lead to an unordered revision list. I think in that case we might just purge newer revisions from the history, and set the selected source repo revision as the latest revision in the history.

Reply:
Nothing prevents users from doing things to create unordered revision lists. It doesn't seem like an issue for plain rsync, but emerge-webrsync is capable of selecting dated snapshots, and nothing stops git sync users from checking out random revisions that can then get recorded in the history.

Reply:
When emerge-webrsync installs a dated snapshot, we can easily wipe previous history and just add the snapshot's revision to the history.

Checkouts of random git revisions are much harder to handle.

Reply:
Maybe we can prevent random git revisions from entering the history by making the GitSync retrieve_head method return an error it's not on a branch that's in sync with a remote tracking branch. Maybe we can use this same mechanism on binhost clients to prevent specifically checked out revisions from entering the history.

Reply:
When a binhost client needs to checkout a specific revision, it can insert that revision into the appropriate location in its revision history, since will have access to git history at that point.

Reply:
I've fixed it to ignore revisions from volatile repos, which seems like it will cover the random git revisions problem.

I still want to handle emerge-webrsync snapshots.

Reply:
Fixed `emerge-webrsync --revert` to do this:
```bash
emaint revisions --purgerepos="${repo_name}"
```